### PR TITLE
Allow RPC yield to be a continuation.

### PR DESCRIPTION
**Pull request by rec**
Reviewers:
@mtrippled @JoelKatz @HowardHinnant 

(Mark and David because they need this facility - Howard for C++ checks...)

### CODE REVIEW

**Comment by rec** [comment]
Bounced circleci!

File: src/ripple/rpc/README.md [hunk 1]
```diff
@@ -1,3 +1,72 @@
-# RPC
+# How to use RPC coroutines.
```
Comment (mtrippled) [inline src/ripple/rpc/README.md]: Titles and section headers don't end with periods, and each word is capitalized except for articles.

**Comment by JoelKatz** [comment]
Unless I'm misreading the code, it seems that if you use coroutines, then the initial execution is not dispatched through the job queue. This breaks pretty much everything in our existing RPC implementation.

**Comment by rec** [comment]
Oh, you mentioned this on informal too.

I believe that we are doing exactly the same as we always did.  I believe that the first step of the coroutine does nothing, just sets it all up and then passes back control and that the payload code is still run on the JobQueue.

Is there an assert I can make to ensure that I am, indeed, running on the `JobQueue` even from the very first goaround?

**Comment by rec** [comment]
All right!

I prepended a new change that allows you to tell if the current thread is running on the `JobQueue`, added an assert, a comment and tested it by running debug and turning on coroutines... and also by removing the initial yield and running and seeing it fail.

File: src/ripple/rpc/Yield.h [hunk 2]
```diff
@@ -27,19 +28,26 @@
...
 
-/** Yield is a generic placeholder for a function that yields control of
-    execution - perhaps to another coroutine.
+/** See the README.md in this directory for more information about how
+    the RPC yield mechanism works.
+ */
 
-    When code calls Yield, it might block for an indeterminate period of time.
+/** Callback: do something and eventually return. */
+using Callback = std::function <void ()>;
```
Comment (mtrippled) [inline src/ripple/rpc/Yield.h]: Where is the original state from where the coroutine eventually yielded? Does the coroutine need to keep the state that it later needs in its object?

**Comment by mtrippled** [comment]
Is the functionality that this PR replaces used by anything currently? Meaning, if this is merged as-is, will anything in the existing code risk behavior change?

**Comment by JoelKatz** [comment]
@rec I'll retest. The reason I suspected this wasn't working correctly is that enabling this functionality caused my server_info output to no longer show the job producing the server_info output. If it was dispatched on the JobQueue, I should see that job running.

**Comment by rec** [comment]
I commented on "the unused `getJob`" but this thread is already long.

It will actually be used by me in code I have "in the other window" regarding https://ripplelabs.atlassian.net/browse/RIPD-909.

**Comment by rec** [comment]
I prefer `boost::optional <Job&>` simply because of the impossibility of undefined behavior regarding uninitialized pointers - and secondarily because the information conveyed is very clear ("might be empty, is owned by someone else").

Since we do this extremely rarely (when I put it in, there are only two functions that actually use `Job&`) and never in a loop the additional boolean is a tiny price to pay.

File: src/ripple/rpc/Yield.h (hunk 2 above)
Comment (rec) [inline src/ripple/rpc/Yield.h]: The state of all local variables are preserved through boost context magic.  You don't need to do anything special - the code looks sequential to you:

    auto foo = expensiveComputation();
    suspend (myContinuation);  // run on jobqueue, wait for condition, etc.
    makeResponse (foo);   // foo is still good.

**Comment by vinniefalco** [comment]
You get undefined behavior if you attempt to dereference a `boost::optional` - same as a pointer. `boost::optional<T&>` offers no benefits over `T*` adding only costs, hence the style rule.

**Comment by rec** [comment]
David: I did do some ninja rebases at the beginning there (basically removing cruft and once fixing a screwup) so you might well have seen something incorrect....

Mark:  as written, we could check this in and the current default codepath (which doesn't use coroutines) would be completely unchanged.  

We could then turn on in a configuration file the coroutines, and separately, yielding during full ledger responses, and then run it on a test machine.

**Comment by rec** [comment]
Indeed, I should make an entire "mutability run" at JobQueue for this, and make the mutex mutable.  brb...

File: src/ripple/server/impl/ServerHandlerImp.h [hunk 3]
```diff
@@ -40,6 +40,7 @@ class ServerHandlerImp
     JobQueue& m_jobQueue;
     NetworkOPs& m_networkOPs;
     std::unique_ptr<HTTP::Server> m_server;
+    RPC::Continuation m_continuation;
```
Comment (vinniefalco) [inline src/ripple/server/impl/ServerHandlerImp.h]: `continuation_` is the new new

**Comment by HowardHinnant** [comment]
One of the reasons `std::experimental::optional<T&>` was made ill-formed is because of the ambiguity of the semantics of assignment of these types.  Should it rebind the reference or assign thru?  Rather than choosing one or the other, and educating clients about that choice, the committee decided to instead make `std::experimental::optional<T&>` a compile-time error.  `boost` chose to rebind.

**Comment by HowardHinnant** [comment]
:+1:

**Comment by mtrippled** [comment]
:+1:

**Comment by rec** [comment]
@JoelKatz ?

This isn't yet changing any behavior.  Just like before, it only goes on when you enable it in the config...

**Comment by rec** [comment]
I just rebased this down for easier reading...

**Comment by JoelKatz** [comment]
It seems to actually work!
:+1:

**Comment by JoelKatz** [comment]
This bit of code re-dispatches the request perfectly:

```
static void doCallback (Job&, RPC::Callback const& c)
{
    c();
}

static void doSuspend (RPC::Callback const& c)
{
    getApp().getJobQueue().addJob (jtCLIENT, "Resume",
        std::bind (&doCallback, std::placeholders::_1, c));
}

Json::Value doServerInfo (RPC::Context& context)
{
    context.suspend (doSuspend);
    ...
````

**Comment by nbougalis** [comment]
I'm seeing thumbs up, so I'm assuming this has passed.

**Pull request closed without merging**